app/src/main/java/com/example/habitsmasher/UserValidator.java

```diff
@@ -0,0 +1,153 @@
+package com.example.habitsmasher;
+
+import android.util.Patterns;
+import android.widget.EditText;
+
+import androidx.core.util.PatternsCompat;
+
+public class UserValidator {
+    private static final String USERNAME_IS_REQUIRED_MESSAGE = "Username is required!";
+    private static final String USERNAME_LENGTH_MUST_BE_LESS_THAN_16_MESSAGE = "Username length must be less than 16!";
+    private static final String EMAIL_IS_REQUIRED_MESSAGE = "Email is required!";
+    private static final String PASSWORD_IS_REQUIRED_MESSAGE = "Password is required!";
+    private static final String INVALID_EMAIL_FORMAT_MESSAGE = "Invalid email format!";
+    private static final String PASSWORD_LENGTH_MUST_BE_GREATER_THAN_6_MESSAGE = "Password length must be greater than 6!";
+    private static final String PASSWORD_LENGTH_MUST_BE_LESS_THAN_16_MESSAGE = "Password length must be less than 16!";
+
+    private EditText _usernameInput;
+    private EditText _emailInput;
+    private EditText _passwordInput;
+
+    String _username;
+    String _email;
+    String _password;
```

Comment:
Just wondering, why are the EditText passed into the validator instead of the string value contained within them?

Reply:
I see. Thanks for explaining!

Reply:
I think I get it, he uses them to set the error messages when not properly formatted. I believe this is in response to the issues with toast messages and detecting them.